transportable-udfs-annotation-processor/src/main/java/com/linkedin/transport/processor/TransportProcessor.java

```diff
@@ -0,0 +1,190 @@
+    try {
+      processImpl(roundEnv);
+    } catch (Exception e) {
+      // We don't allow exceptions of any kind to propagate to the compiler
+      StringWriter writer = new StringWriter();
+      e.printStackTrace(new PrintWriter(writer));
+      fatalError(writer.toString());
+    }
+    // Universal processors should return false since other processor can be potentially acting on the same element
+    return false;
+  }
+
+  private void processImpl(RoundEnvironment roundEnv) {
+    if (roundEnv.processingOver()) {
+      generateUDFPropertiesFile();
+    } else {
+      processElements(roundEnv.getRootElements());
+    }
+  }
+
+  private void processElements(Set<? extends Element> elements) {
+    for (Element element : elements) {
+      // Check if the element is a non-abstract class which extends the StdUDF class
+      if (element.getKind().equals(ElementKind.CLASS) && !element.getModifiers().contains(Modifier.ABSTRACT)
+          && _types.isAssignable(element.asType(), _stdUDFClassType)) {
+        processUDFClass((TypeElement) element);
+      }
+    }
+  }
+
+  /**
+   * Finds the {@link TopLevelStdUDF} for the given {@link StdUDF} class and adds it to the list of discovered UDFs
+   */
+  private void processUDFClass(TypeElement udfClassElement) {
+    log(String.format("Processing UDF Class: %s", udfClassElement.getQualifiedName()));
+
+    TypeElement closestTopLevelStdUDFInterface = getClosestTopLevelStdUDFInterface(udfClassElement);
+    if (closestTopLevelStdUDFInterface != null) {
+      _udfProperties.addUDF(closestTopLevelStdUDFInterface.getQualifiedName().toString(),
+          udfClassElement.getQualifiedName().toString());
+    }
+  }
+
+  /**
+   * Returns a {@link TypeElement} for the closest (in the type hierarchy) class/interface which implements the
+   * {@link TopLevelStdUDF} interface for a given {@link StdUDF} class
+   */
+  private TypeElement getClosestTopLevelStdUDFInterface(TypeElement udfClassElement) {
+    // If the UDF class directly implements TopLevelStdUDF, return the UDF class itself
+    List<? extends TypeMirror> implementedInterfaces = udfClassElement.getInterfaces();
+    for (TypeMirror iface : implementedInterfaces) {
+      if (_types.isSameType(iface, _topLevelStdUDFInterfaceType)) {
+        return udfClassElement;
+      }
+    }
+
+    // Check if any of the interfaces/superclass of the UDF class implement TopLevelStdUDF
+    TypeMirror superClass = udfClassElement.getSuperclass();
+    boolean superClassImplementsTopLevelStdUDFInterface = _types.isAssignable(superClass, _topLevelStdUDFInterfaceType);
+    LinkedList<TypeMirror> candidateInterfaces = new LinkedList<>();
+    for (TypeMirror iface : implementedInterfaces) {
+      if (_types.isAssignable(iface, _topLevelStdUDFInterfaceType)) {
+        candidateInterfaces.add(iface);
+      }
+    }
+
+    if (candidateInterfaces.size() > 1 || (candidateInterfaces.size() == 1
```

Comment:
`candidateInterfaces.size() >= 1` ?

Reply:
Nope. The `superClassImplementsTopLevelStdUDFInterface` only applies to `candidateInterfaces.size() == 1`.